src/odemis/driver/smaract.py

```diff
@@ -3173,3 +2961,1175 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        ret_val = c_int64()
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_i64(self._id, c_uint32(encoded_key), byref(ret_val), c_size_t(0))
+        return ret_val.value
+
+    def GetProperty_s(self, property_key, idx0=0, idx1=0):
+        """
+        property_key (int32): property key symbol
+        idx0 (int): index value, meaning depends on property key
+        idx1 (int): index value, meaning depends on property key
+        returns (str): the value
+        """
+        ret_val = create_string_buffer(SA_SI_STRING_MAX_LENGTH)
+        encoded_key = self.core.SA_SI_EPK(property_key, idx0, idx1)
+        self.core.SA_SI_GetProperty_s(self._id, c_uint32(encoded_key), ret_val, byref(c_size_t(len(ret_val))))
+        return ret_val.value.decode("latin1")
+
+    def GetValue_f64(self, channel, data_source_idx):
+        """
+        property_key (int32): property key symbol
+        data_source_idx:
+        returns (float) the value
+        """
+        ret_val = c_double()
+        self.core.SA_SI_GetValue_f64(self._id, channel, data_source_idx, byref(ret_val))
+        return ret_val.value
+
+    def _load_configuration(self):
+        """
+        Load the configuration of the device. The loaded parameters are
+        * fiber length
+        * extension fiber length
+        * number of active channels
+        * sensor head types
+        * minimum and maximum working distance
+        These configuration parameters need to be set during the installation of the system.
+        Additionally, the parameters from the previously saved referencing procedure are loaded.
+        This function is blocking.
+        """
+        # Loading the configuration will cause the state to become temporarily unstable. Once the process is
+        # finished, the state will be stable again.
+        self.EnableEventNotification(SA_PS_STABLE_STATE_CHANGED_EVENT)
+        self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_RESULT_LOAD_PROP, SA_SI_ENABLED)
+        self._wait_for_event(SA_SI_ENABLED, timeout=600)  # this can take a long time
+        self.DisableEventNotification(SA_PS_STABLE_STATE_CHANGED_EVENT)
+
+    def _save_configuration(self):
+        """
+        Saves the configuration parameters (cf _load_configuration) and internal parameters of the
+        referencing procedure. This function is blocking.
+        """
+        self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_RESULT_SAVE_PROP, SA_SI_ENABLED)
+        # There is no documentation on loading/saving the configuration in the manual, even though
+        # the functions are available in the header file. It is not clear how much time saving the
+        # parameters takes. It can be assumed that it's fast --> wait 1 s.
+        time.sleep(1)
+
+    def _wait_for_event(self, end_state, timeout=float("inf")):
+        """
+        Blocks until event is triggered or timeout.
+        end_state (int): state of event variable to wait for
```

Comment:
It's a little odd to wait only for the state, without check the event type. Pretty much all the states are worth either 0 or 1, so you might end up matching the wrong event (which has a state with the same constant value). Maybe also add a event type?

Reply:
I set it up, so we're only ever waiting for one event.  There was a problem with the event type, I think sometimes the hardware returned event type 0 even though it was a completely different event.